a.go

```diff
@@ -37,6 +37,49 @@ func (a *A) Context() context.Context {
 	return a.ctx
 }
 
+// WithContext returns a shallow copy of a with its context changed
+// to ctx. The provided ctx must be non-nil.
+func (a *A) WithContext(ctx context.Context) *A {
+	if ctx == nil {
+		panic("nil context")
+	}
+
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	result := &A{
+		ctx:     ctx,
+		name:    a.name,
+		output:  a.output,
+		logger:  a.logger,
+		failed:  a.failed,
+		skipped: a.skipped,
+	}
```

Comment:
The issues you encountered with cleanup would be the same for all other mutex protected fields such as `failed` and `skip`.

For instance calling `a.FailNow()` on the returned instance would not cause the original to be marked as failed.

Reply:
Does the same mutex has to be used for `failed`, `skipped` and `cleanups`? If not then maybe we could change the type for `failed` and `skipped` to `*atomic.Bool`? I am not sure if using a pointer to atomic is even good/possible/safe.

Reply:
It seems to me that this is not safe, so look at my solution - it seems to me that it is clearer and safer.

Reply:
@maniak89, I just wanted to show that we could use pointers to atomic (it is safe): https://go.dev/play/p/4FmvNRZfrhB

And I see that `sync.Mutex` as a pointer is also used: https://stackoverflow.com/a/67786582

However, I do not want to go back and forth. I can always refactor it later. This is an implementation detail. I may want to make benchmarks to see what is more efficient in regular uses cases.

**My guess it that for most users  would not ever call `WithContext`. Your proposal should be more efficient in this scenario.**

Reply:
Actually, I think there needs to be pointers.. Otherwise the "failed" and "skipped" will not get propagated to derived `a` which seems not what a user may expect.